panic with generics

related fixes in go-critics:
https://github.com/go-critic/go-critic/pull/1237/files
https://github.com/go-critic/go-critic/pull/1247/files

code:
```
func Map[T, R any](s []T, f func(T) R) []R {
	result := make([]R, len(s))
	for i, v := range s {
		result[i] = f(v)
	}

	return result
}
```

rule:

```
func largeloopcopy(m dsl.Matcher) {
	m.Match(
		`for $_, $v := range $_ { $*_ }`,
	).
		Where(m["v"].Type.Size > 512).
		Report(`loop copies large value each iteration`)
}
```

Stacktrace:
```
panic: /usr/local/Cellar/go/1.19/libexec/src/go/types/sizes.go:181: assertion failed

goroutine 57 [running]:
go/types.assert(0x78?)
        /usr/local/Cellar/go/1.19/libexec/src/go/types/errors.go:27 +0x5b
go/types.(*StdSizes).Sizeof(0xc00002c2b0, {0x14a1c78?, 0xc000308630})
        /usr/local/Cellar/go/1.19/libexec/src/go/types/sizes.go:181 +0x32e
github.com/quasilyte/go-ruleguard/ruleguard.makeTypeSizeConstFilter.func1(0xc0000f57f8)
       go/pkg/mod/github.com/quasilyte/go-ruleguard@v0.3.18/ruleguard/filters.go:445 +0x17c
github.com/quasilyte/go-ruleguard/ruleguard.(*rulesRunner).handleMatch(0xc0000f5680, {0xc0004b4370, 0xc, 0xc0000b0f18, {0xc0003aa031, 0x26}, {0x0, 0x0}, {0x0, 0x0}, ...}, ...)
       go/pkg/mod/github.com/quasilyte/go-ruleguard@v0.3.18/ruleguard/runner.go:352 +0x130
github.com/quasilyte/go-ruleguard/ruleguard.(*rulesRunner).runRules.func1({{0x14a1840, 0xc00019a1e0}, {0xc0000eab00, 0x1, 0x8}})
       go/pkg/mod/github.com/quasilyte/go-ruleguard@v0.3.18/ruleguard/runner.go:252 +0xc5
github.com/quasilyte/gogrep.(*matcher).MatchNode(0xc0004a8540, 0xc0000f56f0, {0x14a1840?, 0xc00019a1e0?}, 0xc0003c34e0)
       go/pkg/mod/github.com/quasilyte/gogrep@v0.0.0-20220828223005-86e4605de09f/match.go:85 +0x4e3
github.com/quasilyte/gogrep.(*Pattern).MatchNode(...)
       go/pkg/mod/github.com/quasilyte/gogrep@v0.0.0-20220828223005-86e4605de09f/gogrep.go:90
github.com/quasilyte/go-ruleguard/ruleguard.(*rulesRunner).runRules(0xc0000f5680?, {0x14a1840?, 0xc00019a1e0?}, 0x23?)
       go/pkg/mod/github.com/quasilyte/go-ruleguard@v0.3.18/ruleguard/runner.go:251 +0x14b
github.com/quasilyte/go-ruleguard/ruleguard.(*rulesRunner).run.func1({0x14a1840?, 0xc00019a1e0?}, 0x1?)
       go/pkg/mod/github.com/quasilyte/go-ruleguard@v0.3.18/ruleguard/runner.go:166 +0x33
github.com/quasilyte/go-ruleguard/ruleguard.(*astWalker).walk(0xc0003c39a0, {0x14a1840?, 0xc00019a1e0?})
       go/pkg/mod/github.com/quasilyte/go-ruleguard@v0.3.18/ruleguard/ast_walker.go:293 +0xb1d
github.com/quasilyte/go-ruleguard/ruleguard.(*astWalker).walkStmtList(0x14a12c8?, {0xc0001fe300?, 0x3, 0x96e8078?})
       go/pkg/mod/github.com/quasilyte/go-ruleguard@v0.3.18/ruleguard/ast_walker.go:37 +0x85
github.com/quasilyte/go-ruleguard/ruleguard.(*astWalker).walk(0xc0003c39a0, {0x14a12c8?, 0xc000191b00?})
       go/pkg/mod/github.com/quasilyte/go-ruleguard@v0.3.18/ruleguard/ast_walker.go:218 +0x181e
github.com/quasilyte/go-ruleguard/ruleguard.(*astWalker).walk(0xc0003c39a0, {0x14a1598?, 0xc000191b30?})
       go/pkg/mod/github.com/quasilyte/go-ruleguard@v0.3.18/ruleguard/ast_walker.go:360 +0x7d9
github.com/quasilyte/go-ruleguard/ruleguard.(*astWalker).walkDeclList(0xc0003c39a0?, {0xc0001e6100?, 0x5, 0xc0003c3920?})
       go/pkg/mod/github.com/quasilyte/go-ruleguard@v0.3.18/ruleguard/ast_walker.go:43 +0x85
github.com/quasilyte/go-ruleguard/ruleguard.(*astWalker).walk(0xc0003c39a0, {0x14a1548?, 0xc0001e6180?})
       go/pkg/mod/github.com/quasilyte/go-ruleguard@v0.3.18/ruleguard/ast_walker.go:367 +0x135e
github.com/quasilyte/go-ruleguard/ruleguard.(*astWalker).Walk(...)
       go/pkg/mod/github.com/quasilyte/go-ruleguard@v0.3.18/ruleguard/ast_walker.go:20
github.com/quasilyte/go-ruleguard/ruleguard.(*rulesRunner).run(0xc0000f5680, 0xc0001e6180)
       go/pkg/mod/github.com/quasilyte/go-ruleguard@v0.3.18/ruleguard/runner.go:165 +0x1c9
github.com/quasilyte/go-ruleguard/ruleguard.(*engine).Run(0x0?, 0x10?, 0x17a7f18?, 0x10?)
       go/pkg/mod/github.com/quasilyte/go-ruleguard@v0.3.18/ruleguard/engine.go:130 +0x45
github.com/quasilyte/go-ruleguard/ruleguard.(*Engine).Run(...)
       go/pkg/mod/github.com/quasilyte/go-ruleguard@v0.3.18/ruleguard/ruleguard.go:74
github.com/quasilyte/go-ruleguard/analyzer.runAnalyzer(0xc0001f9380)
       go/pkg/mod/github.com/quasilyte/go-ruleguard@v0.3.18/analyzer/analyzer.go:139 +0x2a5
golang.org/x/tools/go/analysis/internal/checker.(*action).execOnce(0xc0001c6500)
       go/pkg/mod/golang.org/x/tools@v0.1.12/go/analysis/internal/checker/checker.go:731 +0xa30
sync.(*Once).doSlow(0x1365769?, 0xc0001b81c0?)
        /usr/local/Cellar/go/1.19/libexec/src/sync/once.go:74 +0xc2
sync.(*Once).Do(...)
        /usr/local/Cellar/go/1.19/libexec/src/sync/once.go:65
golang.org/x/tools/go/analysis/internal/checker.(*action).exec(0xc0000103c0?)
       go/pkg/mod/golang.org/x/tools@v0.1.12/go/analysis/internal/checker/checker.go:619 +0x3d
golang.org/x/tools/go/analysis/internal/checker.execAll.func1(0x0?)
       go/pkg/mod/golang.org/x/tools@v0.1.12/go/analysis/internal/checker/checker.go:607 +0x25
created by golang.org/x/tools/go/analysis/internal/checker.execAll
       go/pkg/mod/golang.org/x/tools@v0.1.12/go/analysis/internal/checker/checker.go:613 +0x165


Process finished with the exit code 2


```

--- a/analyzer/testdata/src/generics/rules.go
+++ b/analyzer/testdata/src/generics/rules.go
@@ -10,3 +10,11 @@ func externalErrorReassign(m dsl.Matcher) {
 		Where(m["err"].Type.Is(`error`) && m["pkg"].Object.Is(`PkgName`)).
 		Report(`suspicious reassigment of error from another package`)
 }
+
+func largeLoopCopy(m dsl.Matcher) {
+	m.Match(
+		`for $_, $v := range $_ { $*_ }`,
+	).
+		Where(m["v"].Type.Size > 512).
+		Report(`loop copies large value each iteration`)
+}

--- a/analyzer/testdata/src/generics/target.go
+++ b/analyzer/testdata/src/generics/target.go
@@ -7,3 +7,11 @@ type myType[T any] struct {
 func (m *myType[T]) set(v T) {
 	m.value = v
 }
+
+func Map[T, R any](s []T, f func(T) R) []R {
+	result := make([]R, len(s))
+	for i, v := range s {
+		result[i] = f(v)
+	}
+	return result
+}

--- a/dsl/dsl.go
+++ b/dsl/dsl.go
@@ -222,6 +222,9 @@ func (SinkType) Is(typ string) bool { return boolResult }
 // ExprType describes a type of a matcher expr.
 type ExprType struct {
 	// Size represents expression type size in bytes.
+	//
+	// For expressions of unknown size, like type params in generics,
+	// any filter using this operand will fail.
 	Size int
 }
 

--- a/ruleguard/filters.go
+++ b/ruleguard/filters.go
@@ -436,12 +436,18 @@ func makeTypeSizeConstFilter(src, varname string, op token.Token, rhsValue const
 		if list, ok := params.subNode(varname).(gogrep.ExprSlice); ok {
 			return exprListFilterApply(src, list, func(x ast.Expr) bool {
 				typ := params.typeofNode(x)
+				if isTypeParam(typ) {
+					return false
+				}
 				lhsValue := constant.MakeInt64(params.ctx.Sizes.Sizeof(typ))
 				return constant.Compare(lhsValue, op, rhsValue)
 			})
 		}
 
 		typ := params.typeofNode(params.subExpr(varname))
+		if isTypeParam(typ) {
+			return filterFailure(src)
+		}
 		lhsValue := constant.MakeInt64(params.ctx.Sizes.Sizeof(typ))
 		if constant.Compare(lhsValue, op, rhsValue) {
 			return filterSuccess
@@ -453,8 +459,11 @@ func makeTypeSizeConstFilter(src, varname string, op token.Token, rhsValue const
 func makeTypeSizeFilter(src, varname string, op token.Token, rhsVarname string) filterFunc {
 	return func(params *filterParams) matchFilterResult {
 		lhsTyp := params.typeofNode(params.subExpr(varname))
-		lhsValue := constant.MakeInt64(params.ctx.Sizes.Sizeof(lhsTyp))
 		rhsTyp := params.typeofNode(params.subExpr(rhsVarname))
+		if isTypeParam(lhsTyp) || isTypeParam(rhsTyp) {
+			return filterFailure(src)
+		}
+		lhsValue := constant.MakeInt64(params.ctx.Sizes.Sizeof(lhsTyp))
 		rhsValue := constant.MakeInt64(params.ctx.Sizes.Sizeof(rhsTyp))
 		if constant.Compare(lhsValue, op, rhsValue) {
 			return filterSuccess

--- a/ruleguard/utils.go
+++ b/ruleguard/utils.go
@@ -9,6 +9,8 @@ import (
 	"regexp/syntax"
 	"strconv"
 	"strings"
+
+	"golang.org/x/exp/typeparams"
 )
 
 var invalidType = types.Typ[types.Invalid]
@@ -294,4 +296,9 @@ func identOf(e ast.Expr) *ast.Ident {
 	default:
 		return nil
 	}
+}
+
+func isTypeParam(typ types.Type) bool {
+	_, ok := typ.(*typeparams.TypeParam)
+	return ok
 }

